gdal_edit: add -a_coord_epoch option and update doc

dshean opened the pull request:
<!--
IMPORTANT: Do NOT use GitHub to post any questions or support requests!
           They will be closed immediately and ignored.

Make sure that the title of your commit(s) is descriptive. Typically, they
should be formatted as "component/filename: Describe what the commit does (fixes #ticket)",
so that anyone that parses 'git log' immediately knows what a commit is about.
Do not hesitate to provide more context in the longer part of the commit message.

For example:

"GTiff: fix wrong color interpretation with -co ALPHA=YES (fixes #1234)

When -co ALPHA=YES was used, but PHOTOMETRIC was not specified, the ExtraSample
tag was wrongly set to unspecified.
"

The GDAL project requires specific code formatting for C/C++ and Python code.
This is largely automated with the pre-commit tool.
Consult how to [install and set it up](https://gdal.org/development/dev_practices.html#commit-hooks)

More generally, consult [development practices](https://gdal.org/development/dev_practices.html)
-->

## What does this PR do?

Add new options to `gdal_edit.py` that will add, update or remove a user-defined coordinate epoch from the dataset SRS.  Updated documentation.  

## What are related issues/pull requests?

https://lists.osgeo.org/pipermail/gdal-dev/2024-January/058314.html

## Tasklist

 - [x] Make sure code is correctly formatted (cf [pre-commit configuration](https://gdal.org/development/dev_practices.html#commit-hooks))
 - [x] Add test case(s)
 - [x] Add documentation
 - [x] Updated Python API documentation (swig/include/python/docs/)
 - [ ] Review
 - [x] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

dshean commented:
I don't think there are any tests for `gdal_edit.py`.  I tested several sequences and combinations of `-a_srs`, `-a_coord_epoch` and `-unsetepoch` for a sample GeoTiff.

I added the `-unsetepoch` option to maintain consistency with other existing options, like `-unsetnodata`.  It would be simpler to use `-a_coord_epoch 0.0` to unset, as that is what is returned by GetCoordinateEpoch() when the SRS has no epoch defined.  I expect this will evolve as support for epoch-aware CRS transformations evolves.

No changes to Python API (or associated doc).

rouault commented:
Thanks! Tests for gdal_edit are in autotest/pyscripts/test_gdal_edit.py

rouault commented on doc/source/programs/gdal_edit.rst:
```diff
@@ -58,6 +59,18 @@ It works only with raster formats that support update access to existing dataset
+
+    Assign/override the coordinate epoch of the dataset, as a decimal year (e.g., 2021.3).
+
+    .. versionadded:: 3.4
```
```suggestion
    .. versionadded:: 3.9
```

rouault commented on doc/source/programs/gdal_edit.rst:
```diff
@@ -58,6 +59,18 @@ It works only with raster formats that support update access to existing dataset
+
+    Remove the coordinate epoch of the dataset.
+
+    .. versionadded:: 3.4
```
```suggestion
    .. versionadded:: 3.9
```

dshean commented:
OK, I added a simple test and updated the version number to 3.9 in the doc.

I did a rebase instead of merge before pushing changes.  Let me know if I should clean up, or if you can isolate the relevant commit [0d5fa29](https://github.com/OSGeo/gdal/pull/9144/commits/0d5fa293fdc2c3a079c9066fba7f864496fd724c)

rouault commented:
> I did a rebase instead of merge before pushing changes.

actually, you did the reverse ;-) I'd prefer indeed that it is a rebase with a single commit

dshean commented:
Sorry about that - I'm a little rusty. I reset my local branch, squashed the two relevant commits, and did a force push to my remote branch, which I think takes care of the issue.  Should now be a single commit, I think ready to merge.  Thanks!

rouault commented:
> Sorry about that - I'm a little rusty. I reset my local branch, squashed the two relevant commits, and did a force push to my remote branch, which I think takes care of the issue. Should now be a single commit, I think ready to merge. Thanks!

no problem. that looks good. I'm letting now CI run and see if things are in order

dshean commented:
One other thought - I actually like the FlatGeoBuf strategy to include `EPOCH[2016.47]` in WKT2 (https://gdal.org/user/coordinate_epoch.html#flatgeobuf).  Easy to include this when defining custom WKT2 CRS for a particular dataset, and not limited to GeoTiff format.  Any plans to support that approach in PROJ/GDAL?

Also, it might be useful to have a centralized function to convert a datetime/Timestamp to decimal year for the epoch specification.  I have my own implementation of this, and I recall seeing something in the GDAL code, but I don't think it's exposed in the Python API.

rouault commented:
> Any plans to support that approach in PROJ/GDAL?

well PROJ  (9.2 maybe, not sure exactly which version, but "recently") does support the COORDINATEMETADATA[] WKT support, and when this is ingested by GDAL, it feeds the EPOCH[] to the relevant member of OGRSpatialReference

> Also, it might be useful to have a centralized function to convert a datetime/Timestamp to decimal year for the epoch specification. 

I'm not sure about that. I've always seen the epoch being specified as a decimal year. I imagine that someone does the conversion, but it seems this is something generally done "before" PROJ or GDAL, mostly be geodesists

dshean commented:
> I've always seen the epoch being specified as a decimal year. I imagine that someone does the conversion, but it seems this is something generally done "before" PROJ or GDAL, mostly be geodesists

Yes, epoch should always be specified as decimal year, before updating metadata in PROJ/GDAL.  

I'm just thinking about best place for a convenience function to convert the more common (YYYY, MM, DD) date information to a decimal year, since that's typically what is available from filename or metadata for a dataset.  

Most people who think about these issues have an existing capability to do the conversion to decimal year, usually with their own implementation (something like this: https://stackoverflow.com/a/6451892).  

But it presents an additional hurdle for newer users who want to perform epoch-aware CRS transformations, but only have the more common YYYYMMDD timestamp of their dataset from metadata, and don't have an existing utility to convert.

Not suggesting any change is necessary - users can figure this out on their own.  Alternative might be to add an option for a utility like gdal_edit to accept a timestamp (with known formatting) and then do the conversion to decimal year before setting the epoch.  Or maybe the conversion to decimal year could be easily added to Python datetime module or pandas.Timestamp.
